recipes_source/recipes/tuning_guide.py

```diff
@@ -0,0 +1,366 @@
+# multiple-GPUs. In the case of ``DataParallel``, a single CPU core and a single
+# Python process schedules work for multiple GPUs, which results in a
+# significant cost of launching kernels. For more information refer to the
+# `relevant section of CUDA Best Practices <https://pytorch.org/docs/stable/notes/cuda.html#use-nn-parallel-distributeddataparallel-instead-of-multiprocessing-or-nn-dataparallel>`_
+# from PyTorch documentation.
+
+###############################################################################
+# Skip unnecessary all-reduce if training with DistributedDataParallel and gradient accumulation
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# By default
+# `torch.nn.parallel.DistributedDataParallel <https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel>`_
+# executes gradient all-reduce after every backward pass to compute the average
+# gradient over all workers participating in the training. If training uses
+# gradient accumulation over N steps, then all-reduce is not necessary after
+# every training step, it's only required to perform all-reduce after the last
+# call to backward, just before the execution of the optimizer.
+#
+# `DistributedDataParallel` provides
+# `no_sync() <https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel.no_sync>`_
+# context manager which disables gradient all-reduce for particular iteration.
+# `no_sync()` should applied to first ``N-1`` iterations of gradient
+# accumulation, the last iteration should follow the default execution and
+# perform the required gradient all-reduce.
+
+###############################################################################
+# Match the order of layers in constructors with order during the execution if training with DistributedDataParallel
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# `torch.nn.parallel.DistributedDataParallel <https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel>`_
+# uses the order of layers and parameters from constructors to build buckets for
+# `DistributedDataParallel` gradient all-reduce. `DistributedDataParallel`
+# overlaps all-reduce with the backward pass. All-reduce for a particular bucket
+# is asynchronously triggered only when all gradients for parameters in a given
+# bucket are available.
+#
+# To maximize the amount of overlap, the order in constructors should match the
+# order during the execution. If the order doesn't match, then all-reduce for
+# the entire bucket waits for the gradient which is the last to arrive, this may
+# reduce the overlap between backward pass and all-reduce, all-reduce may end up
+# being exposed, which slows down the training.
```

Comment:
This entire section can be removed.  DDP now rebuckets based on a live backward pass.  There's no need to scare readers into a potentially difficult refactor.

Reply:
@mcarilli Is it also true with find_unused_parameters=True? I think if it's True then DDP still follows order from constructors? So either we remove this section or clarify that it's only needed if find_unused is True (if I'm right here).

Reply:
I believe the dynamic rebucketing is orthogonal to find_unused_parameters=True.  I'll ping @mrshenli and @pritamdamania87 (@pritamdamania ?) on pytorch slack for confirmation.

Reply:
As per my understanding rebucketing is done only if `find_unused_parameters=False`. So I think we should update this section to first mention about rebucketing and then mention that if `find_unused_parameters=True`, this option is disabled and we need to worry about the order in that case. We should retain this section for the `find_unused_parameters=True` case.

Reply:
Heh, i was just checking myself.  Yeah that looks right.  https://github.com/pytorch/pytorch/blob/92f8f75c59ce72b2862ad7f32a78a624b62e451d/torch/csrc/distributed/c10d/reducer.cpp#L491-L500

Reply:
I changed that section and said that this optimization is only for find_unused = True, here is a direct link to the preview: https://deploy-preview-1161--pytorch-tutorials-preview.netlify.app/recipes/recipes/tuning_guide.html#match-the-order-of-layers-in-constructors-and-during-the-execution-if-using-distributeddataparallel-find-unused-parameters-true 
@mcarilli does it look ok?

Reply:
Almost perfect, notes inline.